README.md
# Ansible Collection - znerol.podluck

Runs podman pods under systemd.

This collection provides the modules `znerol.podluck.podluck_container` and
`znerol.podluck_podluck_pod` as well as the role
`znerol.podluck.podluck_pod_systemd`. Modules and role are intended to be
combined into playbooks in order to install pods on target systems managed by
systemd.

Most options of `containers.podman.podman_pod` and
`containers.podman.podman_contaier` can be used as is in
`znerol.podluck_podluck_pod` and `znerol.podluck_podluck_container`. This should
reduce the effort for upgraing from `containers.podman` to `znerol.podluck`.

Note that the systemd units installed by this role are designed to always
*create* and *remove* all containers in a pod when *starting* and *stopping*
them respectively. As a result, the systemd units are working well with `podman
auto-update`.

Units and config can be installed both, in the `user` scope
(`~/.config/podluck`, `~/.config/systemd/user`) as well as in `system` scope
(`/etc/podluck`, `/etc/systemd/system`). As a result, this collection supports
rootfull as well as rootless deployments.

## Requirements

* Ansible>=2.10
* Python 3.6 or better
* containers.podman collection

## Installation

Using a `requirements.yml` file:

```yaml
collections:
  - name: znerol.podluck
```

## Reference

### Options for `znerol.podluck_podluck_pod` module

```
= name
        Name of the pod

        type: str

- path
        Path to the pod configuration file. Use the defaults, since
        this is what the `podluck-service.service` systemd unit
        expects.
        Defaults to `/etc/podluck/pod/{{ name }}/systemd-pod-env` when
        module is executed as `root'.
        Defaults to `~/.config/podluck/pod/{{ name }}/systemd-pod-env`
        when module is executed as an unprivileged user.
        [Default: (null)]
        type: str


- state
        `absent' - A podluck configuration file for the pod in the is
        unlinked if it exists.
        `present' - A podluck configuration file for the pod in the is
        created and populated with cli arguments matching the desired
        configuration. Returns `changed: True` if existing
        configuration does not match the current config. The calling
        playbook / role is responsible for restarting the pod if
        necessary.
        (Choices: absent, present)[Default: present]
        type: str
```

* This module supports check mode.
* This module supports all options provided by
  [containers.podman.podman_pod](https://docs.ansible.com/ansible/latest/collections/containers/podman/podman_pod_module.html#ansible-collections-containers-podman-podman-pod-module) with the following
  exceptions: `executable`, `infra_conmon_pidfile`,
  `pod_id_file`, `recreate`, `replace`.

### Options for `znerol.podluck_podluck_container` module


```
= name
        Name of the container

        type: str

- path
        Path to the container configuration file. Use the defaults,
        since this is what the `podluck-service@.service` systemd unit
        expects.
        Defaults to `/etc/podluck/pod/{{ pod }}/systemd-container-{{
        name }}-env` when module is executed as `root'.
        Defaults to `~/.config/podluck/pod/{{ pod }}/systemd-
        container-{{ name }}-env` when module is executed as an
        unprivileged user.
        [Default: (null)]
        type: str

= pod
        Name of a pod created using [znerol.podluck.podluck_pod].

        type: str

- state
        `absent' - A podluck configuration file for the container in
        the specified pod is unlinked if it exists.
        `present' - A podluck configuration file for the container in
        the specified pod is created and populated with cli arguments
        matching the desired configuration. Returns `changed: True` if
        existing configuration does not match the current config. The
        calling playbook / role is responsible for restarting the
        container / pod if necessary.
        (Choices: absent, present)[Default: present]
        type: str
```

* This module supports check mode.
* This module supports all options provided by
  [containers.podman.podman_container](https://docs.ansible.com/ansible/latest/collections/containers/podman/podman_container_module.html#ansible-collections-containers-podman-podman-container-module) with the following
  exceptions: `cgroups`, `cidfile`, `conmon_pidfile`,
  `detach`, `executable`, `generate_systemd`,
  `infra_conmon_pidfile`, `pod_id_file`, `recreate`,
  `replace`.

### Variables for `znerol.podluck.podluck_pod_systemd` role

```yaml
---
# Required: Pod name, also used as the systemd service name. Must be defined
# when role is imported / included.
# podluck_pod_name: no-default

# Optional: State, either `present` or `absent`. Defaults to `present`.
podluck_pod_state: present

# Recommended: Whether systemd units are enabled (`true`) or disabled (`false`).
# No changes if omitted.
# podluck_pod_enabled: default(omit)

# Recommended: List of container names to start / stop when pod is started /
# stopped.
podluck_pod_containers: []

# Optional: List of dependency entries with container and wants keys:
#     podluck_pod_dependencies:
#       -
#         container: container-name
#         wants: other-container-name
#       -
#         container: other-container-name
#         wants: some-additional-container-name
podluck_pod_dependencies: []

# Optional: Systemd scope (either 'system' or 'user')
podluck_systemd_scope: system

# Optional: Path to systemd units. Defaults to /usr/local/lib/systemd/system if
# `podluck_systemd_scope` is `system` and `~/.config/systemd/user` if
# `podluck_systemd_scope` is `user`.
podluck_systemd_unit_dir: "{{ podluck_systemd_unit_dirs[podluck_systemd_scope] }}"

# Optional: Path to systemd config directory. Defaults to /etc/systemd/system if
# `podluck_systemd_scope` is `system` and `~/.config/systemd/user` if
# `podluck_systemd_scope` is `user`.
podluck_systemd_config_dir: "{{ podluck_systemd_config_dirs[podluck_systemd_scope] }}"

# Optional: Path to directory with podluck systemd unit files. Defaults to the
# ones shipped with this role.
podluck_systemd_skel_src: systemd-skel

# Optional: Whether or not to look for podluck systemd unit files on the remote
# machine. Defaults to `false`.
# podluck_systemd_skel_remote_src: default(omit)

# Optional: Owner of copied systemd unit files. Omitted by default.
# podluck_systemd_unit_file_owner: default(omit)

# Optional: Group of copied systemd unit files. Omitted by default.
# podluck_systemd_unit_file_group: default(omit)

# Optional: Mode of copied systemd unit files. Defaults to rw-r--r--
podluck_systemd_unit_file_mode: 0644

# Optional: Owner of systemd unit override directory. Omitted by default.
# podluck_systemd_override_dir_owner: default(omit)

# Optional: Group of systemd unit override directory. Omitted by default.
# podluck_systemd_override_dir_group: default(omit)

# Optional: Mode of systemd unit override directory. Defaults to rwxr-xr-x
podluck_systemd_override_dir_mode: 0755

# Optional: Owner of systemd unit override files. Omitted by default.
# podluck_systemd_override_file_owner: default(omit)

# Optional: Group of systemd unit override files. Omitted by default.
# podluck_systemd_override_file_group: default(omit)

# Optional: Mode of systemd unit override files. Defaults to rw-r--r--
podluck_systemd_override_file_mode: 0644
```

## Example

### Minimal rootless example

```yaml
---
- name: Minimal rootless example
  hosts: localhost
  gather_facts: no
  vars:
    podluck_systemd_scope: user
    podluck_pod_name: minimal.example.com
    podluck_pod_state: present
    podluck_pod_enabled: true
    podluck_pod_containers:
      - hello-world
  tasks:
    - name: Pod configuration present
      znerol.podluck.podluck_pod:
        name: "{{ podluck_pod_name }}"
        state: "{{ podluck_pod_state }}"
        share: net,uts
        mode: 0600

    - name: Container configuration present
      znerol.podluck.podluck_container:
        name: hello-world
        image: docker.io/library/hello-world:latest
        pod: "{{ podluck_pod_name }}"
        state: "{{ podluck_pod_state }}"
        log_driver: journald
        cap_drop: ALL
        mode: 0600

    - name: Systemd configuration present
      import_role:
        name: znerol.podluck.podluck_pod_systemd

    - name: Pod started
      when: podluck_pod_state != 'absent'
      systemd:
        name: "{{ podluck_pod_name }}.service"
        scope: "{{ podluck_systemd_scope }}"
        state: started
```

After running this playbook, the `minimal.example.com` pod should be up and
running along with `hello-world.minimal.example.com` container. Also systemd
units `minimal.example.com.service` as well as `minimal.example.com@hello-world`
are expected to be started. Inspect the results as follows:

```
$ systemctl --user status minimal.example.com.service
● minimal.example.com.service - Podman pod minimal.example.com managed by podluck
[...]

$ systemctl --user status minimal.example.com@hello-world.service
● minimal.example.com@hello-world.service - Podman container hello-world in pod minimal.example.com managed by podluck
[...]

$ journalctl --user --unit minimal.example.com.service
-- Journal begins at Sat 2021-05-22 14:40:55 CEST, ends at Sun 2021-12-19 14:49:14 CET. --
systemd[676]: Starting Podman pod minimal.example.com managed by podluck...
[...]

$ journalctl --user --unit minimal.example.com@hello-world.service
-- Journal begins at Sat 2021-05-22 14:40:55 CEST, ends at Sun 2021-12-19 14:49:14 CET. --
systemd[676]: Starting Podman container hello-world in pod minimal.example.com managed by podluck...
[...]
conmon[20303]:
conmon[20303]: Hello from Docker!
conmon[20303]: This message shows that your installation appears to be working correctly.
conmon[20303]:
conmon[20303]: To generate this message, Docker took the following steps:
conmon[20303]:  1. The Docker client contacted the Docker daemon.
conmon[20303]:  2. The Docker daemon pulled the "hello-world" image from the Docker Hub.
conmon[20303]:     (amd64)
conmon[20303]:  3. The Docker daemon created a new container from that image which runs the
conmon[20303]:     executable that produces the output you are currently reading.
conmon[20303]:  4. The Docker daemon streamed that output to the Docker client, which sent it
conmon[20303]:     to your terminal.
conmon[20303]:
conmon[20303]: To try something more ambitious, you can run an Ubuntu container with:
conmon[20303]:  $ docker run -it ubuntu bash
conmon[20303]:
conmon[20303]: Share images, automate workflows, and more with a free Docker ID:
conmon[20303]:  https://hub.docker.com/
conmon[20303]:
conmon[20303]: For more examples and ideas, visit:
conmon[20303]:  https://docs.docker.com/get-started/
conmon[20303]:
[...]
systemd[676]: minimal.example.com@hello-world.service: Succeeded.
systemd[676]: minimal.example.com@hello-world.service: Consumed 1.344s CPU time.
```

## License

* [GPL-3 or later](https://www.gnu.org/licenses/gpl-3.0.en.html)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.0.1] - 2023-06-15

### Changed
- Collect podman version (#18)


## [2.0.0] - 2022-01-13

### BREAKING CHANGES

- Update unit files from podluck 2.0.0 (#16)
- Use Wants and PartOf for inter-container dependencies (#15)


## [1.0.5] - 2022-01-22

### Changed
- Ensure that systemd unit dir exists (#11)
- Update to unit files from podluck 1.0.1 (#13)


## [1.0.4] - 2022-01-05

### Changed
- Fix brittle container args (#8)


## [1.0.3] - 2021-12-26
- Update systemd units to podluck 1.0.0 (#5)

## [1.0.2] - 2021-12-19
- Add a README.md for the role

## [1.0.1] - 2021-12-19

### Changed
- Fix quotes in README.md
- Fix ci.
- Add GPL license file.

## [1.0.0] - 2021-12-19
### Added
- Initial import

[Unreleased]: https://github.com/znerol/ansible-collection-podluck/compare/v2.0.1...HEAD
[2.0.1]: https://github.com/znerol/ansible-collection-podluck/compare/v2.0.0...v2.0.1
[2.0.0]: https://github.com/znerol/ansible-collection-podluck/compare/v1.0.5...v2.0.0
[1.0.5]: https://github.com/znerol/ansible-collection-podluck/compare/v1.0.4...v1.0.5
[1.0.4]: https://github.com/znerol/ansible-collection-podluck/compare/v1.0.3...v1.0.4
[1.0.3]: https://github.com/znerol/ansible-collection-podluck/compare/v1.0.2...v1.0.3
[1.0.2]: https://github.com/znerol/ansible-collection-podluck/compare/v1.0.1...v1.0.2
[1.0.1]: https://github.com/znerol/ansible-collection-podluck/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/znerol/ansible-collection-podluck/releases/tag/v1.0.0


plugins/modules/podluck_pod.py
#!/usr/bin/python
# Copyright (c) 2021 Lorenz Schori
#
# GNU General Public License v3.0+
# (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

DOCUMENTATION = r'''
module: podluck_pod
author:
  - "Lorenz Schori"
version_added: '1.0.0'
short_description: Configure podman pods managed by systemd
description:
  - Install and remove configuration for Podman pods managed by systemd
requirements:
  - containers.podman
  - Python 3.6 or better
options:
  name:
    description:
      - Name of the pod
    required: True
    type: str
  path:
    description:
      - >-
        Path to the pod configuration file. Use the defaults, since this
        is what the `podluck-service.service` systemd unit expects.
      - >-
        Defaults to `/etc/podluck/pod/{{ name }}/systemd-pod-env`
        when module is executed as I(root).
      - >-
        Defaults to `~/.config/podluck/pod/{{ name }}/systemd-pod-env`
        when module is executed as an unprivileged user.
    type: str
  state:
    description:
      - >-
        I(absent) - A podluck configuration file for the pod in the is unlinked
        if it exists.
      - >-
        I(present) - A podluck configuration file for the pod in the is created
        and populated with cli arguments matching the desired configuration.
        Returns `changed: True` if existing configuration does not match the
        current config. The calling playbook / role is responsible for
        restarting the pod if necessary.
    type: str
    default: present
    choices:
      - absent
      - present
notes:
  - >-
    This module supports check mode.
  - >-
    This module supports all options provided by
    M(containers.podman.podman_pod) with the following exceptions:
    I(executable), I(infra_conmon_pidfile), I(pod_id_file), I(recreate),
    I(replace).
extends_documentation_fragment:
  - files
'''

EXAMPLES = '''
- name: Pod config for demo.example.com pod present
  znerol.podpourri.podman_pod:
    name: demo.example.com
    state: present
    network: example.com
    share: net,uts

- name: Pod config for demo.example.com pod absent
  znerol.podpourri.podman_container:
    name: demo.example.com
    state: absent
'''

import os  # noqa: F402

from ansible.module_utils.basic import AnsibleModule  # noqa: F402
from ansible_collections.containers.podman.plugins.module_utils.podman.podman_pod_lib import ARGUMENTS_SPEC_POD  # noqa: F402,E501
from ansible_collections.containers.podman.plugins.module_utils.podman.podman_pod_lib import PodmanPodModuleParams  # noqa: F402,E501
from ansible_collections.znerol.podluck.plugins.module_utils.podluck.common import config_get_pod_dir  # noqa: F402,E501
from ansible_collections.znerol.podluck.plugins.module_utils.podluck.common import file_ensure_absent  # noqa: F402,E501
from ansible_collections.znerol.podluck.plugins.module_utils.podluck.common import file_ensure_content  # noqa: F402,E501
from ansible_collections.znerol.podluck.plugins.module_utils.podluck.common import generate_env_content  # noqa: F402,E501
from ansible_collections.znerol.podluck.plugins.module_utils.podluck.common import podman_get_version  # noqa: F402,E501


def _generate_pod_config(module):
    b_command = PodmanPodModuleParams(
        'create',
        module.params,
        podman_get_version(module),
        module
    ).construct_command_from_params()

    return {
        'PODLUCK_POD_ARGS': b_command[1:]
    }


def main():
    ignored_pod_args = [
        'executable',
        'infra_conmon_pidfile',
        'pod_id_file',
        'recreate',
        'replace'
        'state',
    ]

    spec = {
        key: value
        for (key, value)
        in ARGUMENTS_SPEC_POD.items()
        if key not in ignored_pod_args
    }
    spec['backup'] = dict(type='bool', default=False)
    spec['state'] = dict(type='str', default="present", choices=[
        'present',
        'absent',
    ])

    module = AnsibleModule(
        argument_spec=spec,
        add_file_common_args=True,
        supports_check_mode=True,
    )

    name = module.params.pop('name')
    default_dest = os.path.join(
        config_get_pod_dir(name),
        'systemd-pod-env'
    )

    module.params.setdefault('path', default_dest)

    result = dict()
    if module.params['state'] == 'present':
        content = generate_env_content(_generate_pod_config(module))
        result = file_ensure_content(module, content)

    elif module.params['state'] == 'absent':
        result = file_ensure_absent(module)

    # Mission complete
    result['msg'] = f'OK: {result["msg"]}' if 'msg' in result else 'OK'
    module.exit_json(**result)


if __name__ == '__main__':
    main()


plugins/module_utils/podluck/common.py
import errno
import getpass
import os
import pwd
import shlex
import tempfile

from ansible.module_utils._text import to_text


def config_get_base_dir():
    """
    Returns path to config dir.

    Either /etc or ~/.config depending whether or not the module is runnig as
    root or as an unprivileged user.
    """
    pwent = pwd.getpwnam(getpass.getuser())
    if (pwent.pw_uid == 0):
        return '/etc'
    else:
        return os.path.join(pwent.pw_dir, '.config')


def config_get_pod_dir(pod_name):
    """
    Returns path to pod config dir.

    Either /etc/podluck/pod/%pod_name or ~/.config/podluck/pod/%pod_name
    depending whether or not the module is running as root or as an
    unprivileged user.
    """
    return os.path.join(config_get_base_dir(), 'podluck', 'pod', pod_name)


def check_file_attrs(module, changed, message, diff):

    file_args = module.load_file_common_arguments(module.params)
    if module.set_fs_attributes_if_different(file_args, False, diff=diff):

        if changed:
            message += " and "
        changed = True
        message += "ownership, perms or SE linux context changed"

    return message, changed


def file_cleanup(path, result=None):
    # cleanup just in case
    if os.path.exists(path):
        try:
            os.remove(path)
        except (IOError, OSError) as e:
            # don't error on possible race conditions, but keep warning
            if result is not None:
                result['warnings'] = [
                    'Unable to remove temp file (%s): %s' % (path, e)]


def generate_env_content(vars):
    lines = []
    for key, args in vars.items():
        value = ' '.join(shlex.quote(to_text(arg)) for arg in args)
        lines.append(f'{key}={value}\n')
    return ''.join(lines)


def file_ensure_content(module, content):
    changed = False
    path_hash = None
    dest_hash = None
    backup = module.params['backup']

    dest = module.params['path']
    result = {
        'path': dest,
        'state': 'present'
    }

    diff = {
      'before_header': dest,
      'before': ''
    }
    if os.path.exists(dest):
        dest_hash = module.sha1(dest)
        with open(dest, 'r') as existing:
            diff['before'] = ''.join(existing.readlines())

    with tempfile.NamedTemporaryFile(
            mode='w', dir=module.tmpdir, delete=False) as envfile:
        envfile.write(content)

    diff['after_header'] = envfile.name
    diff['after'] = content

    path_hash = module.sha1(envfile.name)
    result['checksum'] = path_hash

    if path_hash != dest_hash:
        if not module.check_mode:
            if backup and dest_hash is not None:
                result['backup_file'] = module.backup_local(dest)

            os.makedirs(os.path.dirname(dest), exist_ok=True)
            module.atomic_move(
                envfile.name,
                dest,
                unsafe_writes=module.params['unsafe_writes']
            )

        changed = True
        msg = 'file changed'
    else:
        msg = 'file up-to-date'

    file_cleanup(envfile.name, result)

    if module.check_mode and not os.path.exists(dest):
        result['diff'] = diff
        result['msg'] = msg
        result['changed'] = changed
    else:
        attr_diff = {}
        msg, changed = check_file_attrs(module, changed, msg, attr_diff)

        attr_diff['before_header'] = '%s (file attributes)' % dest
        attr_diff['after_header'] = '%s (file attributes)' % dest

        difflist = [diff, attr_diff]
        result['diff'] = difflist
        result['msg'] = msg
        result['changed'] = changed

    return result


def file_ensure_absent(module):
    dest = module.params['path']
    result = {
        'path': dest,
        'changed': False,
        'state': 'absent'
    }

    if os.path.exists(dest):

        if not module.check_mode:
            try:
                os.unlink(dest)
            except OSError as e:
                if e.errno != errno.ENOENT:  # It may already have been removed
                    raise e

        result.update({
            'changed': True,
            'diff': {
                'before': {
                    'path': dest,
                    'state': 'present',
                },
                'after': {
                    'path': dest,
                    'state': 'absent',
                },
            },
            'state': 'absent'
        })

    return result

def podman_get_version(module):
    # pylint: disable=unused-variable
    rc, out, err = module.run_command(['podman', b'--version'])
    if rc != 0 or not out or "version" not in out:
        module.fail_json(msg="%s run failed!" % 'podman')
    return out.split("version")[1].strip()


plugins/modules/podluck_container.py
#!/usr/bin/python
# Copyright (c) 2021 Lorenz Schori
#
# GNU General Public License v3.0+
# (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

DOCUMENTATION = r'''
module: podluck_container
author:
  - "Lorenz Schori"
version_added: '1.0.0'
short_description: Configure podman containers managed by systemd
description:
  - Install and remove configuration for Podman containers managed by systemd
requirements:
  - containers.podman
  - Python 3.6 or better
options:
  name:
    description:
      - Name of the container
    required: True
    type: str
  pod:
    description:
      - Name of a pod created using M(znerol.podluck.podluck_pod).
    required: True
    type: str
  path:
    description:
      - >-
        Path to the container configuration file. Use the defaults, since this
        is what the `podluck-service@.service` systemd unit expects.
      - >-
        Defaults to
        `/etc/podluck/pod/{{ pod }}/systemd-container-{{ name }}-env` when
        module is executed as I(root).
      - >-
        Defaults to
        `~/.config/podluck/pod/{{ pod }}/systemd-container-{{ name }}-env` when
        module is executed as an unprivileged user.
    type: str
  state:
    description:
      - >-
        I(absent) - A podluck configuration file for the container in the
        specified pod is unlinked if it exists.
      - >-
        I(present) - A podluck configuration file for the container in the
        specified pod is created and populated with cli arguments matching the
        desired configuration. Returns `changed: True` if existing
        configuration does not match the current config. The calling playbook /
        role is responsible for restarting the container / pod if necessary.
    type: str
    default: present
    choices:
      - absent
      - present
notes:
  - >-
    This module supports check mode.
  - >-
    This module supports all options provided by
    M(containers.podman.podman_container) with the following exceptions:
    I(cgroups), I(cidfile), I(conmon_pidfile), I(detach), I(executable),
    I(generate_systemd), I(infra_conmon_pidfile), I(pod_id_file), I(recreate),
    I(replace).
extends_documentation_fragment:
  - files
'''

EXAMPLES = '''
- name: Container config for wildfly in demo.example.com pod present
  znerol.podpourri.podman_container:
    name: wildfly
    pod: demo.example.com
    image: quay.io/bitnami/wildfly:latest
    state: present
    publish:
      - '8080:8080'
      - '9990:9990'
    volumes:
      - wildfly.demo.example.com:/bitnami/wildfly
    mode: 0600

- name: Container config for wildfly in demo.example.com pod absent
  znerol.podpourri.podman_container:
    name: wildfly
    pod: demo.example.com
    state: absent
'''

import os  # noqa: F402

from ansible.module_utils.basic import AnsibleModule  # noqa: F402
from ansible_collections.containers.podman.plugins.module_utils.podman.podman_container_lib import ARGUMENTS_SPEC_CONTAINER  # noqa: F402,E501
from ansible_collections.containers.podman.plugins.module_utils.podman.podman_container_lib import PodmanModuleParams  # noqa: F402,E501
from ansible_collections.znerol.podluck.plugins.module_utils.podluck.common import config_get_pod_dir  # noqa: F402,E501
from ansible_collections.znerol.podluck.plugins.module_utils.podluck.common import file_ensure_absent  # noqa: F402,E501
from ansible_collections.znerol.podluck.plugins.module_utils.podluck.common import file_ensure_content  # noqa: F402,E501
from ansible_collections.znerol.podluck.plugins.module_utils.podluck.common import generate_env_content  # noqa: F402,E501
from ansible_collections.znerol.podluck.plugins.module_utils.podluck.common import podman_get_version  # noqa: F402,E501


class PodluckModuleParams(PodmanModuleParams):
    def construct_args_from_params(self):
        """Create args from given module parameters.
        Returns:
           list -- list of strings for PODLUCK_CONTAINER_ARGS env var
        """
        cmd = []
        all_param_methods = [func for func in dir(self)
                             if callable(getattr(self, func))
                             and func.startswith("addparam")]
        params_set = (i for i in self.params if self.params[i] is not None)
        for param in params_set:
            func_name = "_".join(["addparam", param])
            if func_name in all_param_methods:
                cmd = getattr(self, func_name)(cmd)

        return cmd


def _generate_container_config(module):
    params = {
        'name': 'ignored_name',
        'image': 'ignored_image',
        'command': [],
    }
    params.update(module.params)
    args = PodluckModuleParams(
        'create',
        params,
        podman_get_version(module),
        module
    ).construct_args_from_params()

    result = {
        'PODLUCK_CONTAINER_ARGS': args,
        'PODLUCK_CONTAINER_IMAGE': [module.params['image']],
    }

    if module.params['command']:
        if isinstance(module.params['command'], list):
            cmd = module.params['command']
        else:
            cmd = module.params['command'].split()
        result['PODLUCK_CONTAINER_IMAGE_CMD'] = cmd

    return result


def main():
    ignored_container_args = [
        'cgroups',
        'cidfile',
        'conmon_pidfile',
        'detach',
        'executable',
        'generate_systemd',
        'infra_conmon_pidfile',
        'pod_id_file',
        'recreate',
        'replace',
    ]
    spec = {
        key: value
        for (key, value)
        in ARGUMENTS_SPEC_CONTAINER.items()
        if key not in ignored_container_args
    }
    spec['pod']['required'] = True
    spec['backup'] = dict(type='bool', default=False)
    spec['state'] = dict(type='str', default="present", choices=[
        'present',
        'absent',
    ])

    module = AnsibleModule(
        argument_spec=spec,
        add_file_common_args=True,
        supports_check_mode=True,
    )

    name = module.params.pop('name')
    pod = module.params.pop('pod')
    default_dest = os.path.join(
        config_get_pod_dir(pod),
        f'systemd-container-{name}-env'
    )

    module.params.setdefault('path', default_dest)

    result = dict()
    if module.params['state'] == 'present':
        content = generate_env_content(_generate_container_config(module))
        result = file_ensure_content(module, content)

    elif module.params['state'] == 'absent':
        result = file_ensure_absent(module)

    # Mission complete
    result['msg'] = f'OK: {result["msg"]}' if 'msg' in result else 'OK'
    module.exit_json(**result)


if __name__ == '__main__':
    main()
